File: src/leetCode/c++/Solution297.cpp
//DFS serialize(), deserialize(): O(V) O(V)
class Codec {//Solution297
public:
    string serialize(TreeNode* root) {
        return serDFS(root);
    }

    string serDFS(TreeNode* root){
        if(!root) return "N";
        return to_string(root->val) + "," + serDFS(root->left) + "," + serDFS(root->right);
    }

    TreeNode* deserialize(string data) {
        queue<string> q;//有可能是兩位數以上, 用string
        string temp = "";
        for(char ch : data){
            if(ch == ','){
                q.push(temp);
                temp.clear();
            }else{
                temp.push_back(ch);
            }
        }
        q.push(temp);//放入最後
        return desDFS(q);
    }

    TreeNode* desDFS(queue<string>& q){//要使用&q否則每次遞迴會不一樣
        string str = q.front(); q.pop();
        if(str == "N") return nullptr;

        int val = stoi(str);
        TreeNode* node = new TreeNode(val);
        node->left = desDFS(q);//依順序解決
        node->right = desDFS(q);
        return node;
    }
};

File: src/leetCode/java/Solution297.java
package leetCode.java;

import java.util.*;

//DFS serialize(), deserialize(): O(V) O(V)
class Codec {//Solution297
    public String serialize(TreeNode root) {
        //Pre-order前序遍歷, root, left, right
        return seDFS(root);
    }

    public String seDFS(TreeNode root){
        if(root == null) return "N";
        return root.val + "," + seDFS(root.left) + "," + seDFS(root.right);
    }

    public TreeNode deserialize(String data) {
        String[] arr = data.split(",");
        Queue<String> queue = new LinkedList<>(Arrays.asList(arr));
        return deDFS(queue);
    }

    public TreeNode deDFS(Queue<String> queue){
        String str = queue.poll();
        if("N".equals(str)) return null;

        int num = Integer.parseInt(str);
        TreeNode root = new TreeNode(num);
        root.left = deDFS(queue);
        root.right = deDFS(queue);
        return root;
    }
}
/*
        1
    2       3
  4   5   N   6
 N N N N     N N

Pre-order:
1,2,4,N,N,5,N,N,3,N,6,N,N
*/